### ISSUE TITLE
Long starting time in RPi

### ISSUE THREAD

**Issue by jofemodo** [x42/tuna.lv2#15]
I just updated to the latest repo version and the plugin now takes more than 20 seconds to start and consuming 100% of CPU. After this starting period, it works normally.

Before the update i was running an older precompiled version from the KXStudio repository. This old version, compiled in 2021, started very quickly.

**Comment by jofemodo** [comment]
OK! It seems the problem is due to building with the "MOD=1" option. It runs fine when building without this option.

**Comment by rerdavies** [comment]
Also experiencing this on on Raspberry Pi.  Something to do with fftw taking10 seconds or so to calculate the fftw plan during load or activate (I'm not sure which).

**Comment by x42** [comment]
Interesting. Is this the official binary from https://x42-plugins.com/x42/x42-tuner or a distro build?

I cannot check on a Pi right now, but will schedule it for the weekend.

**Comment by x42** [comment]
There have not been any significant changes to the plugin in the last 5 or 6 years. 
So if there is a change it is likely due to a libfftw update on the target system.

In a quick test, the statically linked binary is not affected. I have just tested this in 64bit RPI4b using Ardour8, the plugin loads instantly and also immediately responds to input.


So.. to get to the bottom of the issue: Do you compile the plugin yourself? What version of libfftw do you use? What compiler flags do you use?

**Comment by rerdavies** [comment]
I installed the plugin from pre-built packages at https://github.com/x42/tuna.lv2 and also built the plugin from source using sources from https://github.com/x42/tuna.lv2 . The package was built using the default built procedure given on the github page. Both versions of the plugin exhibit the problem.  

It takes 22 seconds to instantiate the plugin on my rpi 4B, substantially all of which is spend in the call to 
```
fftwf_plan_r2r_1d (window_size, ft->fft_in, ft->fft_out, FFTW_R2HC, FFTW_MEASURE)
```

The version of fftw3 is 
```
libfftw3-dev/stable,now 3.3.10-1 arm64 [installed]
```
(with the same version used for subpackages. e.g. 
```
libfftw3-single3/stable,now 3.3.10-1 arm64 [installed,automatic]
```
I think that's the runtime package that's relevant. But all other fftw3 packages have the same version.

I have attached the output of the make (see attached `make_output.txt`).

Compile flags are:
```
cc  -Wall -g -Wno-unused-function -DHAVE_LV2_1_8 -g -fomit-frame-pointer -O3 -fno-finite-math-only -DNDEBUG -DVERSION="\"\"" `pkg-config --cflags lv2 fftw3f` -fPIC -fvisibility=hidden `pkg-config --cflags cairo pangocairo pango` -Irobtk/ -DDISPLAY_INTERFACE   -std=c99 \
  -o build/tuna.so src/tuna.c \
  -shared -Wl,-Bstatic -Wl,-Bdynamic -Wl,--as-needed -pthread  -lm `pkg-config --libs lv2 fftw3f` `pkg-config  --libs cairo pangocairo pango` 
```
It is possible that the  fftw3 library has changed. My copy of Raspberry Pi OS took a major update yesterday, updating raspbery pi os to 
```
$ uname -a
Linux raspberrypi 6.12.20+rpt-rpi-v8 #1 SMP PREEMPT Debian 1:6.12.20-1+rpt1~bpo12+1 (2025-03-19) aarch64 GNU/Linux
```
I have attached output from a clean make in case you might find that useful. 

[make_output.txt](https://github.com/user-attachments/files/19713995/make_output.txt)

I do get error output from the make. Specifically, the message 
```
fatal: No names found, cannot describe anything.
fatal: No names found, cannot describe anything.
fatal: No names found, cannot describe anything.
fatal: No names found, cannot describe anything.
Version:  -> 0 0 0 
fatal: No names found, cannot describe anything.
LV2 Version: 0 0
test -d .git -a .gitmodules -a -f Makefile.git && make -f Makefile.git submodule_check
make[1]: Entering directory '/home/robin/src/tuna.lv2'
Synchronizing submodule url for 'robtk'
make[1]: Leaving directory '/home/robin/src/tuna.lv2'
```
The errors are from `git describe --tags`. Could it be because my fork of the project does not yet have any tags? 

Let me know if I can be of further assistance.

**Comment by rerdavies** [comment]
Incidentally, I just tried the apt package `x42-plugins` . The version of tuna.lv2 distributed in that package on Raspberry Pi OS takes 17 seconds to load. 

```
x42-plugins/stable,now 20221119-1 arm64 [installed]
```

**Comment by x42** [comment]
How about the ready-to-run binary from https://x42-plugins.com/x42/x42-tuner  That is what I tested 

currently https://x42-plugins.com/x42/linux/x42-tuner-v0.6.6-1-arm64.tar.gz
or https://x42-plugins.com/x42/linux/x42-tuner-v0.6.6-1-armhf.tar.gz for 32bit

**Comment by rerdavies** [comment]
Yes. Those binaries exhibit the problem when using the arm64 package

**Comment by x42** [comment]
This is strange, because they do open immediately here after a cold boot, running in Ardour8.

It's an older debian 12 system
```
Linux rpi64 6.1.0-13-arm64 #1 SMP Debian 6.1.55-1 (2023-09-29) aarch64 GNU/Linux
```

yet I don't see how that can have any effect. Those binaries are statically linked and don't use any system libraries except for libc and libm.

**Comment by rerdavies** [comment]
Sincere apologies. Must have fluffed something while installing the pre-built binary before.  :-/

I just downloaded and installed the pre-built binary from https://x42-plugins.com/x42/x42-tuner 

The pre-built binary does **_not_** exhibit the problem.

**Comment by x42** [comment]
Those are statically linked against `fftw-3.3.8`, configured with `--enable-single --enable-float --with-our-malloc --disable-mpi` and `CFLAGS="-O3 -fvisibility=hidden -fPIC`.

It's odd that `fftwf_plan_r2r_1d` takes longer with a fftw-3.3.10. Then again perhaps the issue is deeper.  FFTW has a plan cache, to make that thread-safe, and a mutex to prevent concurrent planning.

**Comment by rerdavies** [comment]
Is there a way to get static linkages in the  build from https://github.com/x42/tuna.lv2 ? 

Are you using a private build of fftw-3.3.8? Or is that the version that's installed on your distro?

**Comment by x42** [comment]
It's a custom build, without using any distro packages (except for gcc, libc, bash and gnu/make)

**Comment by rerdavies** [comment]
Unfortunately, I am unable to downgrade my system to use 3.8 (lots of of system dependencies on fftw3 3.10), so I can't confirm for you that the problem is related to an upstream fix in 3.10 .

It's not a hugely pressing issue for me, since the workaround seems to be to use your prebuilt binary, and because high CPU use makes it impractical for me to use the plugin personally. (Also a bit unfortunate that the version disributed in the debian 12 x42-plugins package is broken).  

Feel free to regard this problem report as a courtesy rather than a request for a fix; and feel free to dispose of the issue as you see fit. 

I remain, regardless of disposition, a great admirer of your work.

Best regards,

-R
